src/main/java/mate.academy/lib/Injector.java

```diff
@@ -1,13 +1,74 @@
+import mate.academy.service.FileReaderService;
+import mate.academy.service.ProductParser;
+import mate.academy.service.ProductService;
+import mate.academy.service.impl.FileReaderServiceImpl;
+import mate.academy.service.impl.ProductParserImpl;
+import mate.academy.service.impl.ProductServiceImpl;
+
 public class Injector {
     private static final Injector injector = new Injector();
+    private static final Map<Class<?>, Class<?>> interfaceImplementations = Map.of(
+            ProductService.class, ProductServiceImpl.class,
+            FileReaderService.class, FileReaderServiceImpl.class,
+            ProductParser.class, ProductParserImpl.class);
+
+    private final Map<Class<?>, Object> instances = new HashMap<>();
 
     public static Injector getInjector() {
         return injector;
     }
 
     public Object getInstance(Class<?> interfaceClazz) {
-        return null;
+        Object clazzImplementationInstance = null;
+        Class<?> clazz = findImplementationClass(interfaceClazz);
+        Field[] declaredFields = clazz.getDeclaredFields();
+        for (Field field : declaredFields) {
+            if (field.isAnnotationPresent(Inject.class)) {
+                Object fieldInstance = getInstance(field.getType());
+                clazzImplementationInstance = createNewInstance(clazz);
+                field.setAccessible(true);
+                try {
+                    field.set(clazzImplementationInstance, fieldInstance);
+                } catch (IllegalAccessException e) {
+                    throw new RuntimeException("Can't inject field " + field.getName()
+                            + " of class " + clazz.getName() + "\n", e);
+                }
+            }
+        }
+        if (clazzImplementationInstance == null) {
+            clazzImplementationInstance = createNewInstance(clazz);
+        }
+        return clazzImplementationInstance;
+    }
+
+    private Object createNewInstance(Class<?> clazz) {
+        if (instances.containsKey(clazz)) {
+            return instances.get(clazz);
+        }
+        try {
+            Object newInstance = clazz.getConstructor().newInstance();
+            instances.put(clazz, newInstance);
+            return newInstance;
+        } catch (ReflectiveOperationException e) {
+            throw new RuntimeException("Can't create new instance of class"
+                    + clazz.getName() + "\n", e);
+        }
+    }
+
+    private Class<?> findImplementationClass(Class<?> interfaceClazz) {
+        Class<?> clazzImpl = interfaceClazz;
+        if (interfaceClazz.isInterface() && interfaceImplementations.containsKey(interfaceClazz)) {
+            clazzImpl = interfaceImplementations.get(interfaceClazz);
+        }
+        if (clazzImpl.isAnnotationPresent(Component.class)) {
+            return clazzImpl;
+        }
+        throw new RuntimeException("Can't create instance of Class: " + clazzImpl.getName()
+                + " - is not annotated with @Component");
```

Comment:
let's check on isAnnotationPresent in your getInstance method right after findImplementationClass() method is called